trieste/space.py

```diff
@@ -331,3 +331,140 @@ def __mul__(self, other: Box) -> Box:
+        :raise ValueError (or InvalidArgumentError): If any of the following are true:
+
+            - ``lower``, ``upper`` and ``stepsizes`` have invalid shapes.
+            - ``lower``, ``upper`` and ``stepsizes`` do not have the same floating point type.
+            - ``upper`` is not greater than ``lower`` across all dimensions.
+        """
+
+        super().__init__(lower, upper)
+        tf.debugging.assert_shapes([(lower, ["D"]), (stepsizes, ["D"])])
+        tf.assert_rank(stepsizes, 1)
+
+        tf.debugging.assert_positive(len(stepsizes), message="stepsizes cannot be empty")
+
+        if isinstance(stepsizes, Sequence):
+            self._stepsizes = tf.constant(stepsizes, dtype=tf.float64)
+        else:
+            self._stepsizes = tf.convert_to_tensor(stepsizes)
+
+            tf.debugging.assert_same_float_dtype([self._lower, self._stepsizes])
+
+    def __repr__(self) -> str:
+        """"""
+        return f"OrdinalSearchSpace({self._lower!r}, {self._upper!r}, {self._stepsizes!r})"
+
+    def sample(self, num_samples: int) -> TensorType:
+        """
+        :param num_samples: The number of points to sample from this search space.
+        :return: ``num_samples`` i.i.d. random points, sampled uniformly, and without replacement,
+            from this search space.
+        """
+        samples = super().sample(num_samples)
+        samples = tf.round(samples / self._stepsizes) * self._stepsizes + 0
+
+        return samples
+
+    @property
+    def stepsizes(self) -> TensorType:
+        """the step sizes of the ordinal space"""
+        return self._stepsizes
+
+    def __contains__(self, value: TensorType) -> bool | TensorType:
+        """
+        Return `True` if ``value`` is a member of this search space, else `False`. A point is a
+        member if all of its coordinates lie in the closed intervals bounded by the lower and upper
+        bounds and spaced by some step sizes.
+
+        :param value: A point to check for membership of this :class:`SearchSpace`.
+        :return: `True` if ``value`` is a member of this search space, else `False`. May return a
+            scalar boolean `TensorType` instead of the `bool` itself.
+        :raise ValueError (or InvalidArgumentError): If ``value`` has a different dimensionality
+            from the search space.
+        """
+        if not shapes_equal(value, self._lower):
+            raise ValueError(
+                f"value must have same dimensionality as search space: {self._lower.shape},"
+                f" got shape {value.shape}"
+            )
+
+        return (
+            tf.reduce_all(value >= self._lower)
+            and tf.reduce_all(value <= self._upper)
+            and tf.reduce_all(
+                tf.map_fn(
+                    lambda x: tf.experimental.numpy.isclose(x[0], x[1]),
+                    (value / self._stepsizes, tf.round(value / self._stepsizes)),
```

Comment:
What is this actually doing?

Reply:
A point on the ordinal space should be divisible by its step size, `value % stepsizes == 0`. But, modulo doesn't work well on float. As you know, there some limitation on  [floating point arithmetic](https://docs.python.org/3/tutorial/floatingpoint.html). So I just take `isclose()` between the result of the division with its nearest integer, the floating point error should be a very small numerical difference between them if the value is divisible by stepsizes. 
Not sure yet what will goes wrong with this approach.